### PR TITLE
Add Flags Tips to abseil.io

### PR DESCRIPTION
**Pull request by manshreck**

### CODE REVIEW

File: _posts/2019-05-10-totw-45.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+variable by reading the code. A flag may not only be set at startup, but could
+potentially be changed later through most of our common HTTP interfaces. If you
+run a server in your binary, there is usually no guarantee that the value of
+your flag will remain unchanged from cycle to cycle, nor any notification if
+it were to change, nor any means for looking for such a change.
+
+Even if you ignore the network-based control "feature", it is often hard (and
+sometimes impossible) to reason about flag values, since flags, by design, may
+be overridden at startup. This uncertainty would be OK if we reliably logged
+binary invocations in production, but we don’t: many production flags are not
```
Comment (tituswinters) [inline _posts/2019-05-10-totw-45.md]: This is all about internal production logging, should be dropped or restructured.

"If your production environment logs the invocation of every binary directly, and stores those logs, great. 
Most environments don't work that way."

File: _posts/2019-05-10-totw-45.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+logged anywhere. If your job is started by a shell script, or you have
+binaries that fork, their invocation is completely unlogged and flag usage
+cannot be determined. For library code, this uncertainty is especially
+nefarious: how can you tell when use of a particular feature is actually
+dead? The simple answer is: you can’t.
+
+Flags also make it challenging to put the final stake in dead code. During
+migrations to new backends, you’d *think* that removing legacy code would
+simply be a matter of removing unnecessary build dependencies and issuing
+history’s most satisfying `g4 rm`. You’d be wrong. If your legacy binary
```
Comment (tituswinters) [inline _posts/2019-05-10-totw-45.md]: s/g4/git/

File: _posts/2019-05-10-totw-90.md [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+type: markdown
+order: "090"
+---
+
+Originally posted as TotW #90 on March 19, 2015
+
+*by [Titus Winters](mailto:titus@google.com)
+
+One of the frustrating things about our (mis)use of command-line flags in
+google3 is the difficulty in removing a flag from binary and production
```
Comment (tituswinters) [inline _posts/2019-05-10-totw-90.md]: Drop ref to google3

File: _posts/2019-05-10-totw-90.md [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+servers safely (revisit https://abseil.io/tips/45 for some frustrating
+misuses). The trouble? A binary won’t start if you specify a flag that is
+no longer defined, and thus removal of flags may require coordination
+between C++ code and your running binaries.
```
Comment (tituswinters) [inline _posts/2019-05-10-totw-90.md]:
```suggestion
between C++ code and your job launching scripts and configurations.
```

File: _posts/2019-05-10-totw-90.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+production code based on a binary version). If only there were a better way!
+
+There is. A while back we added a new concept to the C++ command-line flags
+system called "retired flags" ([`ABSL_RETIRED_FLAG`][1]).
+
+A retired flag creates no symbol in C++ (no more `FLAGS_some_flag` global
+variable to rely upon), does not appear in `--help`, but will be accepted when
+specified on the command line (although an `ERROR` will be logged). In this way,
+flags that are no longer used by the code can be "retired", separating C++
+changes from production code changes. Once the configuration changes are all
```
Comment (tituswinters) [inline _posts/2019-05-10-totw-90.md]:
```suggestion
changes from production configuration changes. Once the configuration changes are all
```

File: _posts/2019-05-10-totw-90.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+clear, the retired flag can be removed once and for all.
+
+Retired flags were designed to be used in many situations (including cases
+involving components or other non-atomic commit requirements). A step-by-step
```
Comment (tituswinters) [inline _posts/2019-05-10-totw-90.md]:
```suggestion
involving cross-repo non-atomic commit requirements). A step-by-step
```

File: _posts/2019-05-10-totw-90.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    should become:
+
+    ```c++
+    ABSL_RETIRED_FLAG(type, frobber, "default", "retired");
+    ```
+
+3.  Wait for binary release(s). Once all the jobs in all of your serving cells
+    are invoking the new binary, you can continue.
+
+4.  Remove the flag from your production code. Once searching the relevant production
```
Comment (tituswinters) [inline _posts/2019-05-10-totw-90.md]:
```suggestion
4.  Remove the flag from your production configs. Once searching the relevant production
```

File: _posts/2019-05-10-totw-90.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    code shows no hits for the frobber flag, you can continue.
```
Comment (tituswinters) [inline _posts/2019-05-10-totw-90.md]:
```suggestion
    configurations shows no hits for the frobber flag, you can continue.
```

File: _posts/2019-05-10-totw-90.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+5.  Remove the retired flag.
+
+We’ve had success with very complicated flag removals: the motivating example
+for this effort was the removal of flags defined in s legacy filesystem, which
```
Comment (tituswinters) [inline _posts/2019-05-10-totw-90.md]:
```suggestion
for this effort was the removal of flags defined in a legacy internal filesystem API, which
```

File: _posts/2019-05-10-totw-45.md [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+variable that cannot be statically predicted, whose usage is incompletely
+logged, and which can only be removed from my code with great difficulty." --
+Nobody, Ever*
+
+The common use of flags in production code, especially within libraries, is
+a mistake. Don’t use flags unless it is truly necessary. There, we said it.
+
+Flags are global variables, only worse: you can’t know the value of that
+variable by reading the code. A flag may not only be set at startup, but could
+potentially be changed later through most of our common HTTP interfaces. If you
```
[the remark this answers is not in the document]
Comment (manshreck) [reply _posts/2019-05-10-totw-45.md]: I kept it because I think the flow around long-standing servers is still applicable. PTAL.

**Comment by manshreck** [comment]
Added edits from code review.

**Pull request merged**